### PR TITLE
Task/dragon 01 features

### PR DESCRIPTION
**Pull request by rg911**
# Dragon 01 Features

##### SecretProofTransaction
- Added 'recipient' (unresolved address) field to SecretProofTransaction. 

##### Split `BlockChainHttp` routes into 3 routes
- BlockHttp
   ```
    getBlockByHeight(height: number): Observable<BlockInfo>;
    getBlockTransactions(height: number, queryParams?: QueryParams): Observable<Transaction[]>;
    getBlocksByHeightWithLimit(height: number, limit: number): Observable<BlockInfo[]>;
    getMerkleReceipts(height: number, hash: string): Observable<MerkleProofInfo>;
    getMerkleTransaction(height: number, hash: string): Observable<MerkleProofInfo>;
    getBlockReceipts(height: number): Observable<Statement>;
   ```
- ChainHttp
   ```
   getBlockchainHeight(): Observable<UInt64>;
   getBlockchainScore(): Observable<BlockchainScore>;
   ```
- DiagnosticHttp
   ```
   getDiagnosticStorage(): Observable<BlockchainStorageInfo>;
   getServerInfo(): Observable<ServerInfo>;
   ```

##### New endpoints
- AccountHttp
   ```
   getAccountsNames(accountIds: Address[]): Observable<AccountNames[]>;
   ```
- MosaicHttp
   ```
   getMosaicsNames(mosaicIds: MosaicId[]): Observable<MosaicNames[]>;
   ```
### MODEL
- Receipt - Migrated from previous PR
- ServerInfo
- MerklrPath
- AccountNames
- MosaicNames

##### E2E
- ModifyMultisigAccountTransaction now changed to be announced with `Aggregated` transaction.
- SecretProofTransaction with new `recipient` added.
- Added `remove cosigners test cases` to multisig account so the e2e test can be re-run without error now. 

#### Levy 
- Removed from both models and transactions

Issue: #147 , #152 , #54
Dependant: Lib (https://github.com/nemtech/nem2-library-js/pull/45)

### CODE REVIEW

File: e2e/infrastructure/BlockHttp.spec.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                [NetworkCurrencyMosaic.createAbsolute(1)],
+                PlainMessage.create('test-message'),
+                NetworkType.MIJIN_TEST,
+            );
+            const signedTransaction = transferTransaction.signWith(account);
+
+            listener.confirmed(account.address).subscribe((transaction: Transaction) => {
+
+                chainHeight = transaction.transactionInfo!.height.lower;
+                console.log(`Test chain height: ${chainHeight}`);
```
Comment (evias) [inline e2e/infrastructure/BlockHttp.spec.ts]: why this debug data ?

File: e2e/infrastructure/Listener.spec.ts [hunk 2]
```diff
@@ -266,6 +341,62 @@ describe('Listener', () => {
         });
     });
 
+    describe('ModifyMultisigAccountTransaction - Restore multisig Accounts', () => {
```
Comment (evias) [inline e2e/infrastructure/Listener.spec.ts]: Not sure why I understand the addon of this transaction type in tests ?

Are all transaction types being tested in the `Listener` test ?

File: e2e/infrastructure/MosaicHttp.spec.ts [hunk 3]
```diff
@@ -54,7 +65,7 @@ describe('MosaicHttp', () => {
         after(() => {
             return listener.close();
         });
-        it('standalone', (done) => {
+        it('Announce MosaicDefinitionTransaction', (done) => {
```
Comment (evias) [inline e2e/infrastructure/MosaicHttp.spec.ts]: How is this related to `MosaicHttp` ? Shouldn't it be located in `TransactionHttp` ?

File: src/infrastructure/ChainHttp.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2018 NEM
```
Comment (evias) [inline src/infrastructure/ChainHttp.ts]: year has changed ;)

File: src/infrastructure/MosaicHttp.ts [hunk 5]
```diff
@@ -98,13 +98,35 @@ export class MosaicHttp extends Http implements MosaicRepository {
...
+        const mosaicIdsBody = {
+            mosaicIds: mosaicIds.map((id) => id.toHex()),
+        };
+        return observableFrom(
+            this.mosaicRoutesApi.getMosaicsNames(mosaicIdsBody)).pipe(map((mosaics) => {
+            return mosaics.map((mosaic) => {
+                return new MosaicNames(
+                    new MosaicId(mosaic.mosaicId),
+                    mosaic.names.map((name) => {
+                        new NamespaceId(name);
```
Comment (evias) [inline src/infrastructure/MosaicHttp.ts]: doesn't this obfuscate the name as it will return namespaceId ?

File: src/infrastructure/receipt/CreateReceiptFromDTO.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+};
+
+/**
+ * @param receiptDTO
+ * @param networkType
+ * @returns {Receipt}
+ * @constructor
+ */
+export const CreateReceiptFromDTO = (receiptDTO, networkType): Receipt => {
+    switch (receiptDTO.type) {
```
Comment (evias) [inline src/infrastructure/receipt/CreateReceiptFromDTO.ts]: Can you link with `@see` or `@link` to the catapult-server source code where types of receipts can be found ?

File: src/infrastructure/transaction/CreateTransactionFromDTO.ts [hunk 7]
```diff
@@ -159,10 +160,10 @@ const CreateStandaloneTransactionFromDTO = (transactionDTO, transactionInfo): Tr
             transactionDTO.nonce,
             new MosaicId(transactionDTO.mosaicId),
             new MosaicProperties(
-                new UInt64(transactionDTO.properties[0].value),
-                (new UInt64(transactionDTO.properties[1].value)).compact(),
-                transactionDTO.properties.length === 3 &&  transactionDTO.properties[2].value ?
-                    new UInt64(transactionDTO.properties[2].value) : undefined,
+                new UInt64(transactionDTO.properties[MosaicPropertyIdEnum.MosaicFlags].value),
```
Comment (evias) [inline src/infrastructure/transaction/CreateTransactionFromDTO.ts]: rename `MosaicPropertyIdEnum` to `MosaicPropertyType`  ?

File: src/model/account/AccountNames.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    constructor(
+        /**
+         * Address of the account.
+         */
+        public readonly address: Address,
+        /**
+         * Address linked namespace names
+         */
+        public readonly names: NamespaceId[]) {
```
Comment (evias) [inline src/model/account/AccountNames.ts]: may require a map with `name: id` as it is not possible currently to get the `name` out of a namespace id.

File: src/model/mosaic/MosaicNames.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    constructor(
+        /**
+         * Mosaic Id
+         */
+        public readonly mosaicId: MosaicId,
+        /**
+         * Address linked namespace names
+         */
+        public readonly names: string[]) {
```
Comment (evias) [inline src/model/mosaic/MosaicNames.ts]: Consistency check: AccountNames uses a NamespaceId array, here you are using a string array. 

Both must be consistent and probably should include `names` values not only namespace ids (see AccountNames PR comment).

File: src/model/mosaic/MosaicPropertyIdEnum.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * The mosaic property id means:
+ * 0 - MosaicFlags
+ * 1 - Divisibility
+ * 2 - Duration
+ */
+export enum MosaicPropertyIdEnum {
```
Comment (evias) [inline src/model/mosaic/MosaicPropertyIdEnum.ts]: proposed rename to `MosaicPropertyType`

File: src/model/receipt/ReceiptType.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Receipt type enums.
+ *
+ * @see https://nemtech.github.io/concepts/receipt.html
```
Comment (evias) [inline src/model/receipt/ReceiptType.ts]: please also link catapult-server source code reference if possible.

File: test/infrustructure/receipt/CreateReceiptFromDTO.spec.ts [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                  0,
+                ],
+                unresolved: [4014740460, 2448037180],
+                resolutionEntries: [
+                  {
+                    source: {
+                      primaryId: 1,
+                      secondaryId: 0,
+                    },
+                    resolved: '08a12f89ee5a49f8',
```
Comment (evias) [inline test/infrustructure/receipt/CreateReceiptFromDTO.spec.ts]: how come the format here is different from an address resolution statement ?

File: e2e/infrastructure/BlockHttp.spec.ts (hunk 1 above)
Comment (rg911) [reply e2e/infrastructure/BlockHttp.spec.ts]: Removed

File: e2e/infrastructure/Listener.spec.ts (hunk 2 above)
Comment (rg911) [reply e2e/infrastructure/Listener.spec.ts]: Well, to test the `BondedTransactionAdded`, 'BondedRemoved` and `cosignatureAdded` event, we need to announce some bonded transactions, and also we need a multisig account with cosigners to be able to co-sign. Having the multisig account restored back to normal account, so the test can be re-run multiple times without errorred out

Comment (rg911) [reply e2e/infrastructure/Listener.spec.ts]: Transactions tests are `in transactionHttpSpec`

File: e2e/infrastructure/MosaicHttp.spec.ts (hunk 3 above)
Comment (rg911) [reply e2e/infrastructure/MosaicHttp.spec.ts]: This is to setup the test mosaic before calling the mosaicHttp. So that the test is not rely on static mosaicId. As the Id changes per node / network

File: src/model/mosaic/MosaicNames.ts (hunk 9 above)
Comment (rg911) [reply src/model/mosaic/MosaicNames.ts]: Fixed

File: src/model/account/AccountNames.ts (hunk 8 above)
Comment (rg911) [reply src/model/account/AccountNames.ts]: Fixed

File: src/infrastructure/transaction/CreateTransactionFromDTO.ts (hunk 7 above)
Comment (rg911) [reply src/infrastructure/transaction/CreateTransactionFromDTO.ts]: Fixed

File: src/model/receipt/ReceiptType.ts (hunk 11 above)
Comment (rg911) [reply src/model/receipt/ReceiptType.ts]: Fixed

File: src/model/mosaic/MosaicPropertyIdEnum.ts (hunk 10 above)
Comment (rg911) [reply src/model/mosaic/MosaicPropertyIdEnum.ts]: Fixed

File: test/infrustructure/receipt/CreateReceiptFromDTO.spec.ts (hunk 12 above)
Comment (rg911) [reply test/infrustructure/receipt/CreateReceiptFromDTO.spec.ts]: that schema has been changed to UInt64 recently. It's now fixed

**Comment by rg911** [comment]
> also, couldn't find file `conf/conf.spec` - how come this is not using `conf/network.conf` ?
> 
> Thank you 👍

I have moved all config settings into the network.conf now. so e2e tests are loading config settings directly from the .conf now

File: src/infrastructure/receipt/CreateReceiptFromDTO.ts (hunk 6 above)
Comment (rg911) [reply src/infrastructure/receipt/CreateReceiptFromDTO.ts]: Fixed

File: src/infrastructure/ChainHttp.ts (hunk 4 above)
Comment (rg911) [reply src/infrastructure/ChainHttp.ts]: Fixed

**Comment by rg911** [comment]
New PR 154 created with dragon 2 features
